File: app/routers/stats.py
"""Statistics and insights API router."""

from datetime import date, datetime, timedelta

from fastapi import APIRouter, Depends, Query, Response
from loguru import logger
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.models.agenda import AgendaEvent, RecurrenceSeries
from app.models.family import FamilyMember
from app.models.meals import Meal
from app.models.photos import Photo
from app.models.tasks import Task, TaskList, TaskRecurrenceSeries

router = APIRouter(prefix="/api/stats", tags=["stats"])


@router.get("/")
async def get_statistics(
    response: Response,
    period: str = Query("all", pattern="^(week|month|year|all)$"),
    db: AsyncSession = Depends(get_db),
):
    """
    Get comprehensive statistics across all modules.

    Periods:
    - week: last 7 days
    - month: last 30 days
    - year: last 365 days
    - all: all time
    """
    logger.info("stats.fetch period={}", period)

    # Cache for 5 minutes - statistics are expensive queries
    response.headers["Cache-Control"] = "private, max-age=300"

    # Calculate date range
    today = date.today()
    if period == "week":
        start_date = today - timedelta(days=7)
    elif period == "month":
        start_date = today - timedelta(days=30)
    elif period == "year":
        start_date = today - timedelta(days=365)
    else:  # all
        start_date = date(2000, 1, 1)

    # ── Database counts ───────────────────────────────────────────────
    family_count_result = await db.execute(select(func.count()).select_from(FamilyMember))
    family_count = family_count_result.scalar() or 0

    task_list_count_result = await db.execute(select(func.count()).select_from(TaskList))
    task_list_count = task_list_count_result.scalar() or 0

    task_count_result = await db.execute(select(func.count()).select_from(Task))
    task_count = task_count_result.scalar() or 0

    event_count_result = await db.execute(select(func.count()).select_from(AgendaEvent))
    event_count = event_count_result.scalar() or 0

    meal_count_result = await db.execute(select(func.count()).select_from(Meal))
    meal_count = meal_count_result.scalar() or 0

    photo_count_result = await db.execute(select(func.count()).select_from(Photo))
    photo_count = photo_count_result.scalar() or 0

    task_series_count_result = await db.execute(select(func.count()).select_from(TaskRecurrenceSeries))
    task_series_count = task_series_count_result.scalar() or 0

    event_series_count_result = await db.execute(select(func.count()).select_from(RecurrenceSeries))
    event_series_count = event_series_count_result.scalar() or 0

    # ── Task completion stats per member ──────────────────────────────
    # Get all completed tasks in period with member associations
    task_completion_query = (
        select(
            FamilyMember.id,
            FamilyMember.name,
            FamilyMember.avatar,
            func.count(Task.id).label("completed_tasks"),
        )
        .join(Task.members)
        .where(Task.done == True)  # noqa: E712
        .where(Task.created_at >= datetime.combine(start_date, datetime.min.time()))
        .group_by(FamilyMember.id, FamilyMember.name, FamilyMember.avatar)
        .order_by(func.count(Task.id).desc())
    )
    task_completion_result = await db.execute(task_completion_query)
    task_completions = [
        {"member_id": row.id, "name": row.name, "avatar": row.avatar, "count": row.completed_tasks}
        for row in task_completion_result.all()
    ]

    # ── Cooking frequency per member ──────────────────────────────────
    cooking_query = (
        select(
            FamilyMember.id,
            FamilyMember.name,
            FamilyMember.avatar,
            func.count(Meal.id).label("meal_count"),
        )
        .join(Meal, Meal.cook_member_id == FamilyMember.id)
        .where(Meal.date >= start_date)
        .group_by(FamilyMember.id, FamilyMember.name, FamilyMember.avatar)
        .order_by(func.count(Meal.id).desc())
    )
    cooking_result = await db.execute(cooking_query)
    cooking_stats = [
        {"member_id": row.id, "name": row.name, "avatar": row.avatar, "count": row.meal_count}
        for row in cooking_result.all()
    ]

    # ── Most common meals ─────────────────────────────────────────────
    meal_frequency_query = (
        select(Meal.name, func.count(Meal.id).label("count"))
        .where(Meal.date >= start_date)
        .group_by(Meal.name)
        .order_by(func.count(Meal.id).desc())
        .limit(10)
    )
    meal_frequency_result = await db.execute(meal_frequency_query)
    top_meals = [{"name": row.name, "count": row.count} for row in meal_frequency_result.all()]

    # ── Agenda activity (events per week) ─────────────────────────────
    # Calculate number of events in period
    event_count_period_result = await db.execute(
        select(func.count())
        .select_from(AgendaEvent)
        .where(AgendaEvent.start_time >= datetime.combine(start_date, datetime.min.time()))
    )
    events_in_period = event_count_period_result.scalar() or 0

    # Calculate weeks in period
    days_in_period = (today - start_date).days
    weeks_in_period = max(days_in_period / 7, 1)
    events_per_week = round(events_in_period / weeks_in_period, 1)

    # ── Task completion rate ──────────────────────────────────────────
    total_tasks_result = await db.execute(
        select(func.count())
        .select_from(Task)
        .where(Task.created_at >= datetime.combine(start_date, datetime.min.time()))
    )
    total_tasks_in_period = total_tasks_result.scalar() or 0

    completed_tasks_result = await db.execute(
        select(func.count())
        .select_from(Task)
        .where(Task.created_at >= datetime.combine(start_date, datetime.min.time()))
        .where(Task.done == True)  # noqa: E712
    )
    completed_tasks_in_period = completed_tasks_result.scalar() or 0

    completion_rate = round(
        (completed_tasks_in_period / total_tasks_in_period * 100) if total_tasks_in_period > 0 else 0, 1
    )

    logger.info(
        "stats.completed period={} task_completions={} cooking={} meals={} events_per_week={}",
        period,
        len(task_completions),
        len(cooking_stats),
        len(top_meals),
        events_per_week,
    )

    return {
        "period": period,
        "database_counts": {
            "family_members": family_count,
            "task_lists": task_list_count,
            "tasks": task_count,
            "task_series": task_series_count,
            "agenda_events": event_count,
            "event_series": event_series_count,
            "meals": meal_count,
            "photos": photo_count,
        },
        "task_completions": task_completions,
        "cooking_frequency": cooking_stats,
        "top_meals": top_meals,
        "agenda_activity": {"events_per_week": events_per_week, "total_events": events_in_period},
        "task_stats": {
            "total": total_tasks_in_period,
            "completed": completed_tasks_in_period,
            "completion_rate": completion_rate,
        },
    }

File: app/static/js/stats.js
/* global API */
(function () {
  let currentPeriod = 'all';

  const loadingEl = document.getElementById('stats-loading');
  const contentEl = document.getElementById('stats-content');

  // ── Load statistics ───────────────────────────────────────────────
  async function loadStats() {
    try {
      loadingEl.classList.remove('hidden');
      contentEl.classList.add('hidden');

      const stats = await API.get(`/api/stats/?period=${currentPeriod}`);
      if (!stats) return;

      renderStats(stats);

      loadingEl.classList.add('hidden');
      contentEl.classList.remove('hidden');
    } catch (err) {
      console.error('Failed to load stats:', err);
      loadingEl.innerHTML = '<div style="text-align: center; padding: 3rem; color: var(--error);">❌ Fout bij laden van statistieken</div>';
    }
  }

  // ── Render statistics ─────────────────────────────────────────────
  function renderStats(stats) {
    // Database counts
    const dbCounts = stats.database_counts;
    document.getElementById('db-family-count').textContent = dbCounts.family_members;
    document.getElementById('db-tasks-count').textContent = dbCounts.tasks;
    document.getElementById('db-events-count').textContent = dbCounts.agenda_events;
    document.getElementById('db-meals-count').textContent = dbCounts.meals;
    document.getElementById('db-photos-count').textContent = dbCounts.photos;
    document.getElementById('db-series-count').textContent = dbCounts.task_series + dbCounts.event_series;

    // Task completion stats
    const taskStats = stats.task_stats;
    document.getElementById('task-completion-rate').textContent = taskStats.completion_rate + '%';
    document.getElementById('task-completed-count').textContent = taskStats.completed;
    document.getElementById('task-total-count').textContent = taskStats.total;

    // Task completions by member
    const taskCompletionList = document.getElementById('task-completion-list');
    const taskCompletionEmpty = document.getElementById('task-completion-empty');

    if (stats.task_completions.length > 0) {
      taskCompletionList.innerHTML = stats.task_completions.map(member => `
        <div class="stat-member-row">
          <div class="stat-member-info">
            <span class="stat-member-avatar">${member.avatar}</span>
            <span class="stat-member-name">${FP.esc(member.name)}</span>
          </div>
          <div class="stat-member-count">
            <span class="stat-count-badge">${member.count}</span>
            <span class="stat-count-label">taken</span>
          </div>
        </div>
      `).join('');
      taskCompletionList.classList.remove('hidden');
      taskCompletionEmpty.classList.add('hidden');
    } else {
      taskCompletionList.classList.add('hidden');
      taskCompletionEmpty.classList.remove('hidden');
    }

    // Cooking frequency
    const cookingList = document.getElementById('cooking-frequency-list');
    const cookingEmpty = document.getElementById('cooking-frequency-empty');

    if (stats.cooking_frequency.length > 0) {
      cookingList.innerHTML = stats.cooking_frequency.map(member => `
        <div class="stat-member-row">
          <div class="stat-member-info">
            <span class="stat-member-avatar">${member.avatar}</span>
            <span class="stat-member-name">${FP.esc(member.name)}</span>
          </div>
          <div class="stat-member-count">
            <span class="stat-count-badge">${member.count}</span>
            <span class="stat-count-label">maaltijden</span>
          </div>
        </div>
      `).join('');
      cookingList.classList.remove('hidden');
      cookingEmpty.classList.add('hidden');
    } else {
      cookingList.classList.add('hidden');
      cookingEmpty.classList.remove('hidden');
    }

    // Top meals
    const topMealsList = document.getElementById('top-meals-list');
    const topMealsEmpty = document.getElementById('top-meals-empty');

    if (stats.top_meals.length > 0) {
      const maxCount = Math.max(...stats.top_meals.map(m => m.count));
      topMealsList.innerHTML = stats.top_meals.map((meal, index) => {
        const percentage = (meal.count / maxCount) * 100;
        return `
          <div class="stat-meal-row">
            <div class="stat-meal-rank">${index + 1}</div>
            <div class="stat-meal-info">
              <div class="stat-meal-name">${FP.esc(meal.name)}</div>
              <div class="stat-meal-bar">
                <div class="stat-meal-bar-fill" style="width: ${percentage}%"></div>
              </div>
            </div>
            <div class="stat-meal-count">${meal.count}×</div>
          </div>
        `;
      }).join('');
      topMealsList.classList.remove('hidden');
      topMealsEmpty.classList.add('hidden');
    } else {
      topMealsList.classList.add('hidden');
      topMealsEmpty.classList.remove('hidden');
    }

    // Agenda activity
    document.getElementById('agenda-events-per-week').textContent = stats.agenda_activity.events_per_week;
    document.getElementById('agenda-total-events').textContent = stats.agenda_activity.total_events;
  }

  // ── Period filter ─────────────────────────────────────────────────
  document.querySelectorAll('.view-btn[data-period]').forEach(btn => {
    btn.addEventListener('click', () => {
      document.querySelectorAll('.view-btn[data-period]').forEach(b => b.classList.remove('active'));
      btn.classList.add('active');
      currentPeriod = btn.dataset.period;
      loadStats();
    });
  });

  // ── Init ──────────────────────────────────────────────────────────
  loadStats();
})();
